src/components/partnership/partnership.repository.ts

```diff
@@ -265,33 +265,82 @@ export class PartnershipRepository extends DtoRepository(Partnership) {
       .apply(calculateTotalAndPaginateList(Partnership, input));
   }
 
-  async verifyRelationshipEligibility(projectId: ID, partnerId: ID) {
+  async verifyRelationshipEligibility(
+    projectId: ID,
+    partnerId: ID,
+    changeset?: ID
+  ) {
     return (
       (await this.db
         .query()
         .optionalMatch(node('partner', 'Partner', { id: partnerId }))
         .optionalMatch(node('project', 'Project', { id: projectId }))
-        .optionalMatch([
-          node('project'),
-          relation('out', '', 'partnership', { active: true }),
-          node('partnership'),
-          relation('out', '', 'partner', { active: true }),
-          node('partner'),
-        ])
+        .subQuery((sub) =>
+          sub
+            .with('project, partner')
+            .optionalMatch([
+              node('project'),
+              relation('out', '', 'partnership', { active: true }),
+              node('partnership'),
+              relation('out', '', 'partner', { active: true }),
+              node('partner'),
+            ])
+            .return(['partnership'])
+            .apply((q) =>
+              changeset
+                ? q
+                    .union()
+                    .with('project, partner')
+                    .match([node('changeset', 'Changeset', { id: changeset })])
+                    .optionalMatch([
+                      node('project'),
+                      relation('out', '', 'partnership', { active: false }),
+                      node('partnership'),
+                      relation('in', '', 'changeset', { active: true }),
+                      node('changeset'),
+                    ])
+                    .optionalMatch([
+                      node('partnership'),
+                      relation('out', '', 'partner', { active: true }),
+                      node('partner'),
+                    ])
+                    .return(['partnership'])
+                : q
+            )
+        )
         .return(['partner', 'project', 'partnership'])
         .asResult<{ partner?: Node; project?: Node; partnership?: Node }>()
         .first()) ?? {}
     );
   }
 
-  async isFirstPartnership(projectId: ID) {
+  async isFirstPartnership(projectId: ID, changeset?: ID) {
     const result = await this.db
       .query()
-      .match([
-        node('project', 'Project', { id: projectId }),
-        relation('out', '', 'partnership', { active: true }),
-        node('partnership'),
-      ])
+      .subQuery((sub) =>
+        sub
+          .match([
+            node('project', 'Project', { id: projectId }),
+            relation('out', '', 'partnership', { active: true }),
+            node('partnership'),
+          ])
+          .return(['partnership'])
+          .apply((q) =>
+            changeset
+              ? q
+                  .union()
```

Comment:
import needed here too

Reply:
hmm... what should we import?

Reply:
Oh I guess nothing. I wonder if it would be better to match project first before the subquery...